internal/platform/fdset_windows.go

```diff
@@ -0,0 +1,239 @@
+	if f == nil {
+		return nil
+	}
+	return &f.sockets
+}
+
+// Regular returns a WinSockFdSet containing the handles in this FdSet that are regular files.
+func (f *FdSet) Regular() *WinSockFdSet {
+	if f == nil {
+		return nil
+	}
+	return &f.regular
+}
+
+// Pipes returns a WinSockFdSet containing the handles in this FdSet that are pipes.
+func (f *FdSet) Pipes() *WinSockFdSet {
+	if f == nil {
+		return nil
+	}
+	return &f.pipes
+}
+
+// getFdSetFor returns a pointer to the right fd set for the given fd.
+// It checks the type for fd and returns the field for pipes, regular or sockets
+// to simplify code.
+//
+// For instance, if fd is a socket and it must be set if f.pipes, instead
+// of writing:
+//
+//	 if isSocket(fd) {
+//			f.sockets.Set(fd)
+//		}
+//
+// It is possible to write:
+//
+//	f.getFdSetFor(fd).Set(fd)
+func (f *FdSet) getFdSetFor(fd int) *WinSockFdSet {
+	h := syscall.Handle(fd)
+	t, err := syscall.GetFileType(h)
+	if err != nil {
+		return nil
+	}
+	switch t {
+	case syscall.FILE_TYPE_CHAR, syscall.FILE_TYPE_DISK:
+		return &f.regular
+	case syscall.FILE_TYPE_PIPE:
+		if isSocket(h) {
+			return &f.sockets
+		} else {
+			return &f.pipes
+		}
+	default:
+		return nil
+	}
+}
+
+// Set adds the given fd to the set.
+func (f *FdSet) Set(fd int) {
+	if s := f.getFdSetFor(fd); s != nil {
+		s.Set(fd)
+	}
+}
+
+// Clear removes the given fd from the set.
+func (f *FdSet) Clear(fd int) {
+	if s := f.getFdSetFor(fd); s != nil {
+		s.Clear(fd)
+	}
+}
+
+// IsSet returns true when fd is in the set.
+func (f *FdSet) IsSet(fd int) bool {
+	if s := f.getFdSetFor(fd); s != nil {
+		return s.IsSet(fd)
+	}
+	return false
+}
+
+// Copy returns a copy of this FdSet. It returns nil, if the FdSet is nil.
+func (f *FdSet) Copy() *FdSet {
+	if f == nil {
+		return nil
+	}
+	return &FdSet{
+		sockets: f.sockets,
+		pipes:   f.pipes,
+		regular: f.regular,
+	}
+}
+
+// Zero clears the set. It returns 0 if the FdSet is nil.
+func (f *FdSet) Count() int {
+	if f == nil {
+		return 0
+	}
+	return f.sockets.Count() + f.regular.Count() + f.pipes.Count()
+}
+
+// Zero clears the set.
+func (f *FdSet) Zero() {
+	if f == nil {
+		return
+	}
+	f.sockets.Zero()
+	f.regular.Zero()
+	f.pipes.Zero()
+}
+
+// Set adds the given fd to the set.
+func (f *WinSockFdSet) Set(fd int) {
+	if f.count < _FD_SETSIZE {
+		f.handles[f.count] = syscall.Handle(fd)
+		f.count++
+	}
+}
+
+// Clear removes the given fd from the set.
+func (f *WinSockFdSet) Clear(fd int) {
+	h := syscall.Handle(fd)
+	for i := uint64(0); i < f.count; i++ {
+		if f.handles[i] == h {
+			for ; i < f.count-1; i++ {
+				f.handles[i] = f.handles[i+1]
+			}
+			f.count--
+			break
+		}
+	}
+}
+
+// IsSet returns true when fd is in the set.
+func (f *WinSockFdSet) IsSet(fd int) bool {
+	h := syscall.Handle(fd)
+	for i := uint64(0); i < f.count; i++ {
+		if f.handles[i] == h {
+			return true
+		}
+	}
+	return false
+}
+
+// Zero clears the set.
+func (f *WinSockFdSet) Zero() {
+	if f == nil {
```

Comment:
unless we use things as nil (later re-assign the pointer) I would leave out the receiver == nil thing for internal types

Reply:
the receiver can be nil in many places because select accepts fdreads, fdwrites, fdexcepts to be nil. By allowing the receiver to be nil, some error checks can be avoided e.g. `f.Count()` returning 0 when f == nil, or in this case, zeroing out a nil receiver, which is also useful in `select`. To be fair, maybe `select` can be rewritten in a different way so that the nil checks are centralized at the top of the function (maybe replacing nils with empty structs)

Reply:
I think what made me look was not all functions check nil ;) Anyway it is cool to know at least some of it is helpful!